components/script/dom/webglrenderingcontext.rs

```diff
@@ -471,6 +545,99 @@ impl WebGLRenderingContext {
         return Ok(expected_byte_length);
     }
 
+    fn flip_teximage_y(&self,
+                       pixels: Vec<u8>,
+                       internal_format: TexFormat,
+                       data_type: TexDataType,
+                       width: usize,
+                       height: usize) -> Vec<u8> {
+        if !self.texture_unpacking_settings.get().contains(FLIP_Y_AXIS) {
+            return pixels;
+        }
+
+        let cpp = (data_type.element_size() *
+                   internal_format.components() / data_type.components_per_element()) as usize;
+
+        let stride = width * cpp;
+
+        // This should have already been validated.
+        assert!(stride * height <= pixels.len());
+
+        let mut flipped = Vec::<u8>::with_capacity(pixels.len());
+
+        for y in 0..height {
+            let flipped_y = height - 1 - y;
+            flipped.extend_from_slice(&pixels[(flipped_y * stride)..((flipped_y + 1) * stride)]);
+        }
+
+        flipped
+    }
+
+    fn premultiply_pixels(&self,
+                          format: TexFormat,
+                          data_type: TexDataType,
+                          pixels: Vec<u8>) -> Vec<u8> {
```

Comment:
Just as above, this only returns a vector of the same size as `pixels`, right?

If that's the case, probably we should also premultiply in-place, either using `&mut [u8]`, or `mut pixels: Vec<u8>`.

Mind at least mentioning it here (and filing a followup maybe?)